test/lit/passes/heap2local.wast

```diff
@@ -2216,3 +2242,903 @@
+    ;; error on validation.
+    (array.get $array
+      (array.new_default $array
+        (i32.const 0)
+      )
+      (i32.const 0)
+    )
+  )
+
+  ;; CHECK:      (func $array.nested.refinalize.set (type $3)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (block (result nullref)
+  ;; CHECK-NEXT:    (ref.null none)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (i32.const 42)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (unreachable)
+  ;; CHECK-NEXT: )
+  (func $array.nested.refinalize.set
+    ;; As above, but with a set.
+    (array.set $array
+      (array.new_default $array
+        (i32.const 0)
+      )
+      (i32.const 0)
+      (i32.const 42)
+    )
+  )
+
+  ;; CHECK:      (func $array.flowing.type (type $1) (result i32)
+  ;; CHECK-NEXT:  (local $temp (ref $array))
+  ;; CHECK-NEXT:  (local $1 i32)
+  ;; CHECK-NEXT:  (local $2 i32)
+  ;; CHECK-NEXT:  (local $3 i32)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (block (result (ref null $2))
+  ;; CHECK-NEXT:    (local.set $1
+  ;; CHECK-NEXT:     (i32.const 1)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (block (result nullref)
+  ;; CHECK-NEXT:     (local.set $3
+  ;; CHECK-NEXT:      (local.get $1)
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:     (local.set $2
+  ;; CHECK-NEXT:      (local.get $3)
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:     (ref.null none)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (block (result (ref null $2))
+  ;; CHECK-NEXT:    (ref.null none)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (block (result (ref null $2))
+  ;; CHECK-NEXT:    (ref.null none)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (block (result i32)
+  ;; CHECK-NEXT:   (drop
+  ;; CHECK-NEXT:    (ref.null none)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (local.get $2)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $array.flowing.type (result i32)
+    (local $temp (ref $array))
+    ;; The array reference here flows through some places that need to be
+    ;; updated when we optimize. In particular the blocks' types will change.
+    (local.set $temp
+      (array.new $array
+        (i32.const 1)
+        (i32.const 1)
+      )
+    )
+    (drop
+      (block $nullable (result (ref null $array))
+        (local.get $temp)
+      )
+    )
+    (drop
+      (block $non-nullable (result (ref $array))
+        (local.get $temp)
+      )
+    )
+    (array.get $array
+      (local.get $temp)
+      (i32.const 0)
+    )
+  )
+
+  ;; CHECK:      (func $get-i32 (type $1) (result i32)
+  ;; CHECK-NEXT:  (i32.const 1337)
+  ;; CHECK-NEXT: )
+  (func $get-i32 (result i32)
+    ;; Helper for the above.
+    (i32.const 1337)
+  )
+)
+
+;; Arrays with reference values.
+(module
+  ;; CHECK:      (type $array (sub (array (ref null $array))))
+  (type $array (sub (array (ref null $array))))
```

Comment:
Is it important that this type is `sub`? If not, perhaps we can elide that part.

Reply:
Sounds good, I removed that part.